### PR TITLE
Support for dumping on Linux.

### PR DESCRIPTION
**Pull request by kingosticks**
Reads /proc/self/maps to discover the text segment start and size.
Function interposition via LD_PRELOAD mechanism.

~~No Linux support for ASLR, no idea how to get that address.~~

I hope I haven't broken OSX but I've no way to test it.

### CODE REVIEW

**Comment by plietar** [comment]
Thanks a lot!
I think the address in `/proc/self/maps` is already adjusted for ASLR, so there's probably nothing to be done there.

Is this just looking at the first line of the file?

**Comment by kingosticks** [comment]
I'm unsure. The text start address seems to be the standard elf one and doesn't appear to change between runs, to be honest I don't know if it's even meant to change (maybe only the linked library addresses are supposed to be randomised e.g. libc ).

It is just looking at the first line. I'm not sure if it's just luck or if it's a standard that it's always first. But if it's not, do you know a good way to identify it? The first line with execute flags set that matches the current program name maybe?

**Comment by kingosticks** [comment]
This was all pretty interesting to look into:
```
cat /proc/20917/maps
00400000-02bdc000 r-xp 00000000 103:05 3411267                           /usr/share/spotify/spotify
02bdd000-02bdf000 r--p 027dc000 103:05 3411267                           /usr/share/spotify/spotify
02bdf000-02fe8000 rwxp 027de000 103:05 3411267                           /usr/share/spotify/spotify
.....
```

From https://eklitzke.org/position-independent-executables:
> Distributions like Debian and Ubuntu only compile particularly security sensitive binaries (e.g. ssh) as PIEs. (Traditional non-PIE binaries will still use ASLR on Linux, but only for loading dynamic libraries).

0x400000 is GNU ld's default value for the text segment start address so what I am seeing all makes sense.

And you are right, /proc/self/maps does include the ASLR as my quick test shows:
```
$ gcc -o testy testy.c -fpic && ./testy
00400000-00401000 r-xp 00000000 103:05 3939092
$ ./testy
00400000-00401000 r-xp 00000000 103:05 3939092
$ ./testy
00400000-00401000 r-xp 00000000 103:05 3939092
$ gcc -o testy testy.c -fpic -pie && ./testy
55cc3efbb000-55cc3efbc000 r-xp 00000000 103:05 3939092
$ ./testy
556f38f44000-556f38f45000 r-xp 00000000 103:05 3939092
$ ./testy
559fd3d27000-559fd3d28000 r-xp 00000000 103:05 3939092
```

So yeh, seems all good!

File: dump/dump.sh [hunk 1]
```diff
@@ -1,3 +1,5 @@
 #!/usr/bin/env bash
 set -eux
-DYLD_INSERT_LIBRARIES=dump.dylib "$@"
+FILENAME=dump.dylib
+FILEPATH="$(cd "$(dirname "$FILENAME")"; pwd)/$(basename "$FILENAME")"
```
Comment (plietar) [inline dump/dump.sh]: Why did you add these? Given that FILENAME is a constant, this will always be the same value anyway.

**Comment by kingosticks** [comment]
LD_PRELOAD requires the full path, not the relative path.

File: dump/dump.sh (hunk 1 above)
Comment (kingosticks) [reply dump/dump.sh]: LD_PRELOAD requires the full path, not the relative path.

Comment (plietar) [reply dump/dump.sh]: Just use $PWD/$FILENAME then.

`dirname $FILENAME` will always be `.`, and `basename $FILENAME` will always be `$FILENAME`

File: dump/Makefile [hunk 2]
```diff
@@ -1,12 +1,20 @@
 CC=gcc
 CFLAGS=-Wall -Wextra -Wmissing-declarations
 
+UNAME_S:=$(shell uname -s)
+ifeq ($(UNAME_S),Linux)
+	CFLAGS += -D LINUX -fPIC -shared
+endif
+ifeq ($(UNAME_S),Darwin)
+	CFLAGS += -D OSX -dynamiclib
```
Comment (plietar) [inline dump/Makefile]: You can use `__APPLE__` and `__linux__` in `dump.c` instead of defining your own preprocessor flags

File: dump/dump.c [hunk 3]
```diff
@@ -102,11 +109,30 @@ static void patch_shn(void) {
...
-    void *text_start = getsectdata("__TEXT", "__text", &text_size) + aslr_offset;
+    text_start = getsectdata("__TEXT", "__text", &text_size) + aslr_offset;
+#endif
+
+#ifdef LINUX
+    unsigned long long start, end = 0;
+    FILE *mapfile = fopen("/proc/self/maps", "r");
+    if (mapfile) {
+        char line[256];
+        if (fgets(line, sizeof(line), mapfile)) {
```
Comment (plietar) [inline dump/dump.c]: Add a note and a TODO that this only looks at the first line and should be made more robust.
(I don't really care about fixing it now, but I don't want it to be lost).

File: dump/dump.c [hunk 4]
```diff
@@ -102,11 +109,30 @@ static void patch_shn(void) {
...
     printf("ASLR slide: 0x%lx\n", aslr_offset);
 
-    size_t text_size;
-    void *text_start = getsectdata("__TEXT", "__text", &text_size) + aslr_offset;
+    text_start = getsectdata("__TEXT", "__text", &text_size) + aslr_offset;
+#endif
+
+#ifdef LINUX
+    unsigned long long start, end = 0;
+    FILE *mapfile = fopen("/proc/self/maps", "r");
```
Comment (plietar) [inline dump/dump.c]: You can use `assert(mapfile != NULL)` like we do in other places rather than continuing

**Review by plietar — CHANGES_REQUESTED** [review]

File: dump/dump.sh (hunk 1 above)
Comment (kingosticks) [reply dump/dump.sh]: Yeh I guess it's ott in this particular case. I'll push that when I get a chance again or feel free yourself.

**Comment by kingosticks** [comment]
Addressed your comments.

**Comment by plietar** [comment]
Awesome

**Comment by kingosticks** [comment]
Ah ok, now I remember why I had `FILEPATH="$(cd "$(dirname "$FILENAME")"; pwd)/$(basename "$FILENAME")"`, it allowed you to run it from an arbitrary directory which was quite nice. Nevermind.
